### PR TITLE
[DatePicker] Initial commit

### LINKED ISSUE #2398: [RFC] Migrate date pickers from the lab to MUI X
This issue is meant to give a space to discuss moving the date picker components from the lab to MUI X.
I propose that once we assign a full-time product & engineering owner on them, we do the move, and expose the components as:

- `@mui/x-pickers`
- `@mui/x-pickers-pro`

The component is currently a shared responsibility of the core team, where Sebastian has lately been the one spending the most time on them. 

Advantages:
- If we continue with the date range picker inside the Pro plan, it seems simpler to isolate the whole codebase here, to keep the core fully MIT.
- The core has been successful up until now without these date pickers components. Moving them here would allow the core team to focus on the DX and the design dimensions on the core. From what I understand, it's what is ultimately better aligned with the success of the product this team focuses on
- It would allow giving more development bandwidth to the date pickers. It feels like there is a significant backlog of problems that we could focus on, but it's currently hard to balance time between the date picker and the other core components.

Disadvantages:

- One could argue that the last time we moved the components from *mui-org/material-ui-pickers* to *mui-org/material-ui" it was a mess. We ended up having to update a lot of the code of the component. I personally that this was great. It forced us to make the picker components consistent with the other ones. While the same applies here? Yes, likely. I do expect that the migration won't be very smooth. However, what I hope is the opposite outcome as in the previous migration. Meaning, we won't touch the component but it will force us to make sure MUI X has the same development environment as the main repository.
- One could argue that the core should own any component that has a native counterpart. Why? I didn't follow the reasoning so I would leave this to who wants to defend the argument.
Instead, I can argue why the opposite might work better: Bootstrap is successful with few components. Focusing on too many components in the core feels risky, it could prevent the team to focus on what has more impact: the DX for the most common use cases, the design appeal, and the ease of customizations.

### PR DESCRIPTION
**Pull request by flaviendelangle**
Closes #2398

[Doc preview](https://deploy-preview-3451--material-ui-x.netlify.app/x/react-date-pickers/getting-started/)
[Codesandbox pro example](https://codesandbox.io/s/datagrid-v5-quick-start-forked-5kwtg7?file=/src/App.tsx)

## Changes

For the view types, I removed the duplicates and only kept the type with the name of the view component, not the ones with the name of the pickers (`clock` not `time`, `calendar` not `date`, `calendarOrClock` not `dateTime`)

- ~~`ClockView`~~, ~~`TimePickerView`~~, `ClockPickerView` 
- ~~`DatePickerView`~~, `CalendarPickerView`
- ~~`DateTimePickerView`~~, `CalendarOrClockPickerView`

Removing `ClockPickerStandalone` and instead have `ClockPicker` handle the controlled / uncontrolled pattern just like in `CalendarPicker`

- [x] Add `root` slot for `ClockPicker`

- [x] Rename `PickersCalendar` into `DayPicker`

## PRs ready to be picked:
- https://github.com/mui-org/material-ui/pull/30514
- https://github.com/mui-org/material-ui/pull/30749 => DONE
- https://github.com/mui-org/material-ui/pull/30773 => DONE
- https://github.com/mui-org/material-ui/pull/30768 => DONE
- https://github.com/mui-org/material-ui/pull/30738 => DONE
- https://github.com/mui-org/material-ui/pull/29071
- https://github.com/mui-org/material-ui/pull/30786 => DONE
- https://github.com/mui/material-ui/pull/30561 => DONE

### CODE REVIEW

**Comment by siriwatknp** [comment]
👋 The [migration PR](https://github.com/mui/mui-x/pull/3730) has been merged.

Please follow these steps to make sure that the contents are all updated. (Sorry for the inconvenience)

1. pull latest master from upstream to your branch
2. if your PR has changes on the `*.md` or demo files, you are likely to encounter conflict because all of them have been moved to the new folder.
   2.1 revert the change on those markdown files you did
   2.2 pull latest master from upstream (you should not see conflict)
   2.3 make the changes again at `docs/data/material/*`
3. run `yarn docs:api`
    - you might see the changes in `docs/pages/material/api/*` if your PR touches some of the components
    - it is okay if there is no changes

If you are struggle with the steps above, feel free to tag @siriwatknp

**Comment by siriwatknp** [comment]
@flaviendelangle Would all the `/x/api/date-pickers/*` pages included in this PR?

**Comment by flaviendelangle** [comment]
@siriwatknp they should yes

File: docs/data/date-pickers/time-picker/time-picker.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+# Time Picker
+
+<p class="description">Time pickers allow the user to select a single time.</p>
+
+Time pickers allow the user to select a single time (in the hours:minutes format).
+The selected time is indicated by the filled circle at the end of the clock hand.
+
+## Basic usage
+
+The date picker is rendered as a modal dialog on mobile, and a textbox with a popup on desktop.
```
Comment (alexfauquette) [inline docs/data/date-pickers/time-picker/time-picker.md:18]:
```suggestion
The time picker is rendered as a modal dialog on mobile, and a textbox with a popup on desktop.
```

File: docs/data/date-pickers/date-time-picker/date-time-picker.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+materialDesign: https://material.io/components/date-pickers
+---
+
+# Date Time Picker
+
+<p class="description">Combined date & time picker.</p>
+
+This component combines the date & time pickers. It allows the user to select both date and time with the same control.
+
+Note that this component is the [DatePicker](/components/pickers/date-picker/) and [TimePicker](/components/pickers/time-picker/)
```
Comment (alexfauquette) [inline docs/data/date-pickers/date-time-picker/date-time-picker.md:15]: I don't know why, but in the peview: this link redirect to "/material-ui/pickers/date-picker/"

**Review by alexfauquette — COMMENTED** [review]
I did not see any big problem in the component doc.

About the API documentation, scripts should be adapted for two points:

- The "Component name" section is for the DataGrid, and not the date pickers. See for example [calendarPickerSkeleton API](https://deploy-preview-3451--material-ui-x.netlify.app/x/api/date-pickers/calendar-picker-skeleton/)

> Component name
>
> The name `MuiDataGrid` can be used when providing `default props` or `style overrides` in the theme.

On the same idea, all the import examples are made from `@mui/x-data-grid-pro` instead of the correct package

- In the two following pages, the description column is empty which is not the case in the mui.com version
	- https://deploy-preview-3451--material-ui-x.netlify.app/x/api/date-pickers/date-range-picker/
	- https://deploy-preview-3451--material-ui-x.netlify.app/x/api/date-pickers/calendar-picker/
	
Not for this PR: A follow up do don't forget is to add an explanation in the get started about how to deal with the two packages. When reading demos, I was wondering if the `LocalizationProvider` could be imported for `@mui/x-data-picker-pro`

File: docs/data/date-pickers/date-time-picker/date-time-picker.md (hunk 2 above)
Comment (siriwatknp) [reply docs/data/date-pickers/date-time-picker/date-time-picker.md:15]: That's the logic from the core repo. It should be fine when we enable the redirects.

Comment (siriwatknp) [inline docs/data/date-pickers/date-time-picker/date-time-picker.md:15]:
```suggestion
Note that this component is the [DatePicker](/x/react-date-pickers/date-picker/) and [TimePicker](/x/react-date-pickers/time-picker/)
```

Comment (flaviendelangle) [reply docs/data/date-pickers/date-time-picker/date-time-picker.md:15]: Good catch, I did not updated the links in the pages
The pickers docs will never be released on the old structure, so we can change them

Comment (flaviendelangle) [reply docs/data/date-pickers/date-time-picker/date-time-picker.md]: https://github.com/mui/mui-x/pull/3451/commits/7d74bfb0afca8766f3baef95c86264682c37a53b

I think I got them all

File: docs/data/date-pickers/time-picker/time-picker.md (hunk 1 above)
Comment (flaviendelangle) [reply docs/data/date-pickers/time-picker/time-picker.md:18]: This one was there on the core :D 
Fixed

**Comment by flaviendelangle** [comment]
@alexfauquette concerning the references to `DataGridPro` or `x-data-grid-pro` in the api page, I removed some hardcoded values and synced the `ApiPage` file with the core (I can't use it directly because we have the concept of slots which they do not document).

**Comment by alexfauquette** [comment]
I finally found Why date-range-picker description was empty, it's a regexp problem

The current one `/date-range-picker.*.json$/` is valid for files such as 

- 'docs/translations/api-docs/date-pickers/static-date-range-picker.json'
-  'docs/translations/api-docs/date-pickers/mobile-date-range-picker.json',
- 'docs/translations/api-docs/date-pickers/desktop-date-range-picker.json',
- 'docs/translations/api-docs/date-pickers/date-range-picker-day-pt.json',
- 'docs/translations/api-docs/date-pickers/date-range-picker-day-zh.json',
- ...

The `.` needs to be escaped and the lang tag must be restrained to 2 letters to avoid accepting "date-range-picker-day"

A solution can be `/\/date-range-picker(-[a-z]{2})?\.json$/` 
- 'docs/translations/api-docs/date-pickers/date-range-picker.json',
- './date-range-picker.json',
- './date-range-picker-pt.json',
- './date-range-picker-zh.json',
- 'docs/translations/api-docs/date-pickers/date-range-picker-pt.json',
- 'docs/translations/api-docs/date-pickers/date-range-picker-zh.json'

**Comment by flaviendelangle** [comment]
> A solution can be /\/date-range-picker(-[a-z]{2})?\.json$/

Done

**Review by alexfauquette — APPROVED** [review]

**Pull request merged**